modules/javafx.graphics/src/main/java/com/sun/javafx/css/PseudoClassState.java

```diff
@@ -54,35 +54,6 @@ public PseudoClassState() {
         }
     }
 
-    /** {@inheritDoc} */
-    @Override
-    public Object[] toArray() {
-        return toArray(new PseudoClass[size()]);
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public <T> T[] toArray(T[] a) {
```

Comment:
It's definitely a bug, but it would only appear if you had bit sets which required more than 1 long to store their data **and** you called `toArray`.  It would then throw an `ArrayIndexOutOfBoundsException` in this line:

    final long state = getBits()[index];

This is because `index` is used for two unrelated purposes:
- the index into the array of longs
- the index of the entry in the resulting array

If the outer loop ever needs two passes, it again wants to read a `long` using `getBits()[index]` but as `index` was advanced (multiple times) while filling entries in the result array, it won't be the correct index at all.

Anyway, I'm not sure what the value would be of a test.  Since `toArray` is no longer implemented by us but by `AbstractSet`, I'd just be testing the `AbstractSet` code.

Reply:
Okay, thanks for the clarification. 

So if I get it right, the removal of `toArray` doesn't have to do with the bug, but with the fact that with the immutable set it is not longer required?

Reply:
I removed it because my fix requires that `toArray` works correctly.  The easiest way to get a correctly working version is to extend `AbstractSet`, which provides a default implementation that works correctly.  As I think the default implementation is good enough and performs well enough, I saw no reason to fix the broken version.

Reply:
The original ("broken") version has been working fine, and no bugs have been reported so far, and there would be a reason to have a custom implementation instead of the one in `AbstractSet` in the first place. 

I'm not against removing it, but only after we are certain that this implementation is no longer needed. 

Also, have you tried fixing it instead of removing it? If you have, are there any differences when you run the test with one or the other?

Reply:
You could check this yourself if you want.  The `BitSet` class had problems in many places, was largely untested and, to be very honest, should never have passed code review.  It violated the `Set` contract almost everywhere, which is problematic since sets backed by this `BitSet` class were exposed in several places where they could be accessed by users.

A simple test shows the problem (it doesn't throw an exception though, I misread that):

    public static void main(String[] args) {
        for (int i = 0; i < 65; i++) {
            PseudoClassState.getPseudoClass("" + i);
        }

        System.out.println(Arrays.asList(new PseudoClassState(List.of("0", "1", "64")).toArray()));
    }

Prints:

    [0, null, null]

There is no point in fixing the existing code; it won't perform any better, but would require writing additional test cases to verify that an implementation we don't need is doing what we can get for free.

Reply:
Alright, I'll add the test.

As for the other issue: there are quite a few other issues **near** the code I've changed, but in order for a PR to remain focused, and not burdened with changes that don't contribute to the intended change (making it smaller and easier to review), it's best to address these in another PR that focuses specifically on such problems (unused constructors, unused fields, unused methods, etc).

In this specific case, the constructor was not changed or added by me, it's part of private API and it's not causing problems.  Changing it would distract from the goal of the PR, and so is IMHO out of scope.

Reply:
I agree this is out of scope. Once this PR is merged, there are many things that can be done to cleanup the code.